planner/core/rule_join_reorder_dp.go

```diff
@@ -18,32 +18,47 @@ import (
 
 	"github.com/pingcap/parser/ast"
 	"github.com/pingcap/tidb/expression"
-	"github.com/pingcap/tidb/sessionctx"
 )
 
 type joinReorderDPSolver struct {
-	ctx     sessionctx.Context
-	newJoin func(lChild, rChild LogicalPlan, eqConds []*expression.ScalarFunction) LogicalPlan
+	*baseSingleGroupJoinOrderSolver
+	newJoin func(lChild, rChild LogicalPlan, eqConds []*expression.ScalarFunction, otherConds []expression.Expression) LogicalPlan
 }
 
-type joinGroupEdge struct {
+type joinGroupEqEdge struct {
 	nodeIDs []int
 	edge    *expression.ScalarFunction
 }
 
-func (s *joinReorderDPSolver) solve(joinGroup []LogicalPlan, conds []expression.Expression) (LogicalPlan, error) {
-	adjacents := make([][]int, len(joinGroup))
-	totalEdges := make([]joinGroupEdge, 0, len(conds))
-	addEdge := func(node1, node2 int, edgeContent *expression.ScalarFunction) {
-		totalEdges = append(totalEdges, joinGroupEdge{
+type joinGroupNonEqEdge struct {
+	nodeIDs []int
+	idMask  uint
+	expr    expression.Expression
```

Comment:
how about:
- `s/idMask/nodeIDMask/`
- `s/expr/edge/` <- to keep the same variable naming style of `joinGroupEqEdge`

Reply:
What about `content`? If we rename it to `edge`, there'll be call like `edge.edge` in code.